portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -33,42 +35,135 @@
+      String mood = (String) entity.getProperty("mood");
+      long timestamp = (long) entity.getProperty("timestamp");
+      String userFilter = (String) entity.getProperty("filter");
+
+      Comment comment = new Comment(id, text, author, mood, timestamp, userFilter);
       comments.add(comment);
     }
 
-    Gson gson = new Gson();
+    return comments;
+  }
 
-    response.setContentType("application/json;");
-    response.getWriter().println(gson.toJson(comments));
+  /* Create filters for timestap and text of comment */
+  public static Comparator timestampFilter() {
+    Comparator<Comment> byTimestamp = Comparator.comparingLong(Comment::getTimestamp);
+    return byTimestamp;
   }
 
-  @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get the input from the form.
+  public static Comparator textFilter() {
+    Comparator<Comment> byLength = Comparator.comparingInt(Comment::getTextLength);
+    return byLength;
+  }
+
+  /* Get the filter from the last added comment => that is the filter wanted */
+  public static String getLastFilter(List<Comment> comments) {
+    Collections.sort(comments, timestampFilter().reversed());
+    return comments.get(0).getFilter();
+  }
+
+  /* Filter the comments => sort by timestamp or text */
+  public static void sort(List<Comment> comments, String userFilter) {
+    if ("newest".equals(userFilter)) {
+      Collections.sort(comments, timestampFilter().reversed());
+    } else if ("oldest".equals(userFilter)) {
+      Collections.sort(comments, timestampFilter());
+    } else if ("longest".equals(userFilter)) {
+      Collections.sort(comments, textFilter().reversed());
+    }
+  }
+
+  /* Convert the filtered comments to json */
+  public static String toJson(List<Comment> comments) {
+    Gson gson = new Gson();
+    return gson.toJson(comments);
+  }
+
+  /* Get the input from the form, create entity and add to datastore*/
+  public static void constructFromInput(HttpServletRequest request) {
     String userComment = request.getParameter("user-comment");
     String userName = request.getParameter("user-name");
+    String userMood = request.getParameter("mood");
+    long timestamp = System.currentTimeMillis();
+    String userFilter = getFilterChoice(request);
 
     Entity commentEntity = new Entity("Comment");
     commentEntity.setProperty("text", userComment);
     commentEntity.setProperty("author", userName);
+    commentEntity.setProperty("mood", userMood);
+    commentEntity.setProperty("timestamp", timestamp);
+    commentEntity.setProperty("filter", userFilter);
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
+  }
 
-    // Redirect back to the HTML page.
-    response.sendRedirect("/comment.html");
+  /* Get the requested filter and also look for input errors */
+  public static String getFilterChoice(HttpServletRequest request) {
```

Comment:
A small comment on naming: "filter" makes me think of e.g. the `where` statement in SQL, or choosing a few items out of many according to certain criteria. "Sort order" or "sorting order" or "sorting criterion" are terms I have seen in the past.